### ISSUE TITLE
$filter does not return an array if no or only one item matches

### ISSUE THREAD

**Issue by brodo** [jsonata-js/jsonata#218]
I don't know if this is on purpose, but it lead to some problems for me today:

```
// data: [0,0,5,0]
$filter($, function($e) { $e != 0 })
//result: 5, not [5]

// data: [1,2,3,4]
$filter($, function($e) { $e > 5 })
/result: ** no match **
```

At least for the case that there is only one result, I expect there to be an array.

**Comment by wnm3** [comment]
We found a workaround to force creation of an array with a single cell... we used the $append([],value) which returns [value]. So, for example:
$append([],"hello")
returns ["hello"]

**Comment by andrew-coleman** [comment]
This change was caused by this [commit](https://github.com/jsonata-js/jsonata/commit/0fe0774aa0b611e69d29cbecf9c0effef38676cf) which went into v1.5 as a result of issue #93.  It attempted to bring more consistency to the handling of result sequences, and in this case brought the `$filter` function in line with the filter syntax `[...]` and the `$map` function in line with the map operator `.`.  

As with the path syntax, the empty filter operator `[]` should be used to keep singleton sequences as arrays, e.g. `$filter($, function($e) { $e != 0 })[]` should produce `[5]` in your case.  However, I've noticed that this is not working as expected when applied to a function, so I'll fix that in the next maintenance release.

**Comment by deepilla** [comment]
@andrew-coleman I don't quite get the idea behind result sequences. I can see the benefit in a function that takes, say, an array of strings also being able to handle a single string. But what's the advantage of a function like `$filter` sometimes returning an array and sometimes not? It seems like that would make it more difficult to use if anything.

**Comment by andrew-coleman** [comment]
This is now fixed in v1.6.1.

@deepilla I've attempted to explain the rationale of result sequences in a new section of the docs [here](http://docs.jsonata.org/processing#sequences).  I'm still working on improving the documentation, but hopefully this gives the idea.
